compiler/qsc_project/src/project.rs

```diff
@@ -657,16 +657,24 @@ impl PackageGraphSources {
         }
 
         let mut sorted_packages = self.packages.into_iter().collect::<Vec<_>>();
+        let mut cycle_detected = false;
```

Comment:
How does this differentiate between package missing and cycle detected?

Reply:
I will admit, this algorithm is a little terse/loosely commented -- I'll update that.

On line 644, this algorithm filters out any dependencies which have greater than 0 dependencies, since we know they won't be able to be the first element in the build order. So, only dependency-less packages are in the `queue`. Every package should be reachable from this `queue`, as they are all inserted as neighbors in the initial construction of `graph`. The only situation in which a package is not present in `sorted_keys` but _is_ present in `sorted_packages` is if it is not reachable via a path which starts with a package that has zero dependencies. In other words, if all possible build orders that reach this package contain a cycle, then there is no possible package to begin the build order with, as every package on the path has at least one dependency it must compile first.

Reply:
I just documented this more thoroughly in this commit: d8dc9b1a